Envar filter and remove

mgonzalezperna opened the pull request.

mgonzalezperna committed:
Add functions to remove envars from devices.
There is still WIP. Its not tested.

mgonzalezperna committed:
Add functions to start dealing with devices envars

This functions are the backbone of the ones declared on the cli.

mgonzalezperna committed:
Add to cli a command to filter and delete envars.

The command gives you the chance to delete envars filtering by values.
The values to filter can be a list.
Also, the filtering could be defined inclusive or exclusive.

mgonzalezperna committed:
Add the backend to the cli function.

The methods uses Balena API to filter devices and enviroments variables.
It validates the input values to, raising exceptions if some of the
inputs is not valid (environment model invalid, environment variable not
existant, value to filter without matches).

It will also need one more level of fitering for the services model (app
services and device services) because it does not allow to check for
environment variables for a specific service name.

Roger commented on resin_release_tool/cli.py:
```diff
@@ -118,5 +118,50 @@ def releases(releaser, count):
+    try:
+        devices = releaser.get_devices_filtered_by_condition(
+            envar_model, envar_name, envar_value, exclusive_condition)
+    except Exception as exception:
```
this exception is too broad

mgonzalezperna replied:
Fixed.

Roger commented on resin_release_tool/releaser.py:
```diff
@@ -21,6 +21,237 @@ def get_canaries(self):
+        try:
+            self.models.environment_variables.application.remove(envar['id'])
+            results = 'Success!'
+        except:
```
this exception is too broad

mgonzalezperna replied:
Fixed.

Roger commented on resin_release_tool/releaser.py:
```diff
@@ -21,6 +21,237 @@ def get_canaries(self):
+            self.models.environment_variables.\
+                service_environment_variable.remove(envar['id'])
+            results = 'Success!'
+        except:
```
this exception is too broad

mgonzalezperna replied:
Fixed.

Roger commented on resin_release_tool/releaser.py:
```diff
@@ -21,6 +21,237 @@ def get_canaries(self):
+                self.models.environment_variables.\
+                    device_service_environment_variable.remove(envar_id)
+                results['done'] += 1
+            except:
```
this exception is too broad

mgonzalezperna replied:
Fixed

mgonzalezperna committed:
Add BalenaException class handler in remove methods

mgonzalezperna committed:
Exclusive parameter for filtering now is optional.

mgonzalezperna committed:
Major refactor over releaser.
Fixed duplicated code, now all the methods are more cohesive and
shorter.

mgonzalezperna committed:
Fix Exception catching.

mgonzalezperna committed:
Fix CS.

Roger commented on resin_release_tool/cli.py:
```diff
@@ -118,5 +118,50 @@ def releases(releaser, count):
+@click.pass_context
+def filter_and_remove_env_var(ctx, releaser, envar_model, \
+                              envar_name, envar_value, exclusive):
+    """Filter devices by enviroment variables and removes them."""
```
Can you add more documentation here, like that each variable means and help to exclusive?

mgonzalezperna replied:
Sure! On it

mgonzalezperna committed:
Add comments for arguments and options.

mgonzalezperna committed:
Refactor of how to call the function from cli.

To make the tool a little more easy on the user, we refactored how to
pass arguments to the remove function. This should improve declarativity
on the use of the command.

mgonzalezperna committed:
Adapt documentation to new way of passing args.

mgonzalezperna committed:
Refactor to interact with the reviewed releaser.

We needed a major overhauling over the releaser, so, naturally, the cli
has to be adapted to work with the new functions. Now devices returns a
list of dict with device_uuid:envar_key searched, so its easier to pass
the values to the remover function and we save a second call to the
filtering function.

mgonzalezperna committed:
Move the environment constants to a enum class.

mgonzalezperna committed:
Fix of CS of a string.

mgonzalezperna committed:
Major refactor to add service filtering.

The overhaul of the releaser serves to add service filtering for envars,
as well to make the code more simple and concise. Now, the devices are
returned with the envar_id of the envar that makes then fulfill the
filtering condition. This helps us to later remove the envar,
but will also become handy if we eventually would want to update the envars
or do another operations. Also, its easier to get from the SDK the info
necessary to compare and filter (Sadly, Balena SDK suffers of a annoying lack
of abstractions and requieres a lot of 'data massaging' to get the
filtering done right).

mgonzalezperna committed:
Add handlers to params processing on remove method.
